### ISSUE TITLE
Misuses not caught when String flows to constructor call

### ISSUE THREAD

**Issue by enriozuni** [CROSSINGTUD/CryptoAnalysis#164]
In various CryptoGuard projects in issue #134 that were headless tested, misuses were not caught in the following scenario from the analysis. The cases include a particular scenario where a variable is used as parameter in the constructor call of an object being created. I have listed below the projects and the objects from the JCA that were present in these cases.

# CryptoGuard projects and their respective cases
Below are cases that were headless tested. They are grouped according to their project name in CryptoGuard and the JCA object involved.
* __brokencrypto__  _(Cipher)_
  * The following example is [BrokenCryptoABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokencrypto/BrokenCryptoABSCase1.java)
```java
package example.brokencrypto;

import javax.crypto.*;
import java.io.UnsupportedEncodingException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

public class BrokenCryptoABSCase1 {
    Crypto2 crypto;
    public BrokenCryptoABSCase1() throws NoSuchAlgorithmException, NoSuchPaddingException {
        crypto = new Crypto2("DES/ECB/PKCS5Padding");
    }
}

class Crypto2 {
    Cipher cipher;
    String defaultAlgo;
    public Crypto2(String defAlgo) throws NoSuchPaddingException, NoSuchAlgorithmException {
        defaultAlgo = defAlgo;
    }

    public byte[] encrypt(String txt, String passedAlgo) throws UnsupportedEncodingException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException, NoSuchAlgorithmException, NoSuchPaddingException {
        if(passedAlgo.isEmpty()){
            passedAlgo = defaultAlgo;
        }

        KeyGenerator keyGen = KeyGenerator.getInstance(defaultAlgo);
        SecretKey key = keyGen.generateKey();
        Cipher cipher = Cipher.getInstance(defaultAlgo);
        cipher.init(Cipher.ENCRYPT_MODE, key);

        byte [] txtBytes = txt.getBytes();
        return cipher.doFinal(txtBytes);
    }
}
```
No ConstraintError regarding insecure algorithm used in Cipher is reported. Other similar cases in the same project folder include: [BrokenCryptoABSCase2](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokencrypto/BrokenCryptoABSCase2.java), [BrokenCryptoABSCase3](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokencrypto/BrokenCryptoABSCase3.java), [BrokenCryptoABSCase4](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokencrypto/BrokenCryptoABSCase4.java), [BrokenCryptoABSCase5](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokencrypto/BrokenCryptoABSCase5.java).

* __brokenhash__  _(MessageDigest)_
Other similar cases in this project folder include:
  * [BrokenHashABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokenhash/BrokenHashABSCase1.java)
  * [BrokenHashABSCase2](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokenhash/BrokenHashABSCase2.java)
  * [BrokenHashABSCase3](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokenhash/BrokenHashABSCase3.java)
  * [BrokenHashABSCase4](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/brokenhash/BrokenHashABSCase4.java)

* __ecbcrypto__  _(Cipher)_
Other similar cases in this project folder include:
  * [EcbInSymmCryptoABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/ecbcrypto/EcbInSymmCryptoABSCase1.java)

* __pbeiteration__  _(PBEParameterSpec)_
Other similar cases in this project folder include:
  * [LessThan1000IterationPBEABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/pbeiteration/LessThan1000IterationPBEABSCase1.java)

* __predictablecryptographickey__  _(SecretKeySpec)_
Other similar cases in this project folder include:
  * [PredictableCryptographicKeyABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/predictablecryptographickey/PredictableCryptographicKeyABSCase1.java)

* __predictablekeystorepassword__  _(KeyStore)_
Other similar cases in this project folder include:
  * [PredictableKeyStorePasswordABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/predictablekeystorepassword/PredictableKeyStorePasswordABSCase1.java)

* __predictablepbepassword__  _(PBEKeySpec)_
Other similar cases in this project folder include:
  * [PredictablePBEPasswordABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/predictablepbepassword/PredictablePBEPasswordABSCase1.java)

* __predictableseeds__  _(SecureRandom)_
Other similar cases in this project folder include:
  * [PredictableSeedsABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/predictableseeds/PredictableSeedsABSCase1.java)
  * [PredictableSeedsABSCase2](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/predictableseeds/PredictableSeedsABSCase2.java)

* __staticinitializationvector__  _(IvParameterSpec)_
Other similar cases in this project folder include:
  * [StaticInitializationVectorABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/staticinitializationvector/StaticInitializationVectorABSCase1.java)

* __staticsalts__  _(PBEParameterSpec)_
Other similar cases in this project folder include:
  * [StaticSaltsABSCase1](https://github.com/CryptoGuardOSS/cryptoapi-bench/blob/master/src/main/java/org/cryptoapi/bench/staticsalts/StaticSaltsABSCase1.java)

**Comment by johspaeth** [comment]
Thanks for reporting. Have you tried calling the `encrypt` method? It seems as noone ever calls the method.

Please add the line `crypto.encrypt("")`  to the constructor `BrokenCryptoABSCase1()` and try again.

**Comment by enriozuni** [comment]
@johspaeth yes I have tried and the analysis works correctly in that case

**Comment by johspaeth** [comment]
I close the issue, if there is no call to encrypt, the error cannot occur at runtime.

**Comment by enriozuni** [comment]
@johspaeth OK, got it. Gonna update the headless test cases as well
